state/virtualhostkeys.go

```diff
@@ -0,0 +1,109 @@
+package state
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"encoding/pem"
+
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/core/virtualhostkeys"
+	"github.com/juju/mgo/v3"
+	"github.com/juju/mgo/v3/txn"
+)
+
+// VirtualHostKey represents the state of a virtual host key.
+type VirtualHostKey struct {
+	st  *State
+	doc virtualHostKeyDoc
+}
+
+type virtualHostKeyDoc struct {
+	DocId   string `bson:"_id"`
+	Id      string `bson:"id"`
+	HostKey []byte `bson:"hostkey"`
+}
+
+// Id returns the virtual host key's ID.
+func (s *VirtualHostKey) Id() string {
+	return s.doc.Id
+}
+
+// HostKey returns the virtual host key.
+func (s *VirtualHostKey) HostKey() []byte {
+	return s.doc.HostKey
+}
+
+func newVirtualHostKeyDoc(hostKeyID string) (virtualHostKeyDoc, error) {
+	_, privateKey, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return virtualHostKeyDoc{}, errors.Annotate(err, "failed to generate ED25519 key")
+	}
+
+	pemKey, err := gossh.MarshalPrivateKey(privateKey, "")
+	if err != nil {
+		return virtualHostKeyDoc{}, errors.Annotate(err, "failed to marshal private key")
+	}
+	return virtualHostKeyDoc{
+		// Unsure on docID vs Id.
+		DocId:   hostKeyID,
+		Id:      hostKeyID,
+		HostKey: pem.EncodeToMemory(pemKey),
+	}, nil
+}
+
+func newMachineVirtualHostKeysOps(machineID string) ([]txn.Op, error) {
+	hostKeyID := virtualhostkeys.MachineHostKeyID(machineID)
+	doc, err := newVirtualHostKeyDoc(hostKeyID)
+	if err != nil {
+		return nil, err
+	}
+	return []txn.Op{{
+		C:      virtualHostKeysC,
+		Id:     doc.DocId,
+		Assert: txn.DocMissing,
+		Insert: doc,
+	}}, nil
+}
+
+func newUnitVirtualHostKeysOps(unitName string) ([]txn.Op, error) {
+	hostKeyID := virtualhostkeys.UnitHostKeyID(unitName)
+	doc, err := newVirtualHostKeyDoc(hostKeyID)
+	if err != nil {
+		return nil, err
+	}
+	return []txn.Op{{
+		C:      virtualHostKeysC,
+		Id:     doc.DocId,
+		Assert: txn.DocMissing,
+		Insert: doc,
+	}}, nil
+}
+
+func removeVirtualHostKeysOps(docID string) []txn.Op {
+	return []txn.Op{{
+		C:      virtualHostKeysC,
+		Id:     docID,
+		Remove: true,
+	}}
+}
+
+// VirtualHostKey retrieves host key by ID.
+func (st *State) VirtualHostKey(id string) (*VirtualHostKey, error) {
+	vhkeys, closer := st.db().GetCollection(virtualHostKeysC)
+	defer closer()
+
+	doc := virtualHostKeyDoc{}
+	err := vhkeys.FindId(id).One(&doc)
+	if err == mgo.ErrNotFound {
+		return nil, errors.NotFoundf("virtual host key %q", id)
+	}
+	if err != nil {
+		return nil, errors.Annotatef(err, "cannot get virtual host key %q", id)
```

Comment:
The fact that you can't is implied by the error. These tend to read  better as they go up the stack if we just state the attempted action, in this case "getting virtual host key".